sky/packages/sky/lib/rendering/object.dart

```diff
@@ -30,50 +30,92 @@ class ParentData {
 
-  sky.PictureRecorder _recorder;
+  // Don't keep a reference to the PaintingContext.canvas, since it
+  // can change dynamically after any call to this object's methods.
 
   PaintingContext(Offset offest, Size size) {
     _startRecording(offest, size);
   }
 
-  PaintingContext.forTesting(this.canvas);
+  PaintingCanvas _canvas;
+  PaintingCanvas get canvas => _canvas;
+
+  PictureLayer _layer;
+  PictureLayer get layer => _layer;
+
+  sky.PictureRecorder _recorder;
+
+  PaintingContext.forTesting(this._canvas);
 
   void _startRecording(Offset offset, Size size) {
     assert(_layer == null);
     assert(_recorder == null);
-    assert(canvas == null);
+    assert(_canvas == null);
     _layer = new PictureLayer(offset: offset, size: size);
     _recorder = new sky.PictureRecorder();
-    canvas = new PaintingCanvas(_recorder, Point.origin & size);
+    _canvas = new PaintingCanvas(_recorder, Point.origin & size);
   }
 
   void endRecording() {
     assert(_layer != null);
     assert(_recorder != null);
-    assert(canvas != null);
-    canvas = null;
+    assert(_canvas != null);
     _layer.picture = _recorder.endRecording();
-    _recorder = null;
     _layer = null;
+    _recorder = null;
+    _canvas = null;
+  }
+
+  bool debugCanPaintChild() {
+    // You need to use layers if you are applying transforms, clips,
+    // or similar, to a child. To do so, use the paintChildWith*()
+    // methods below.
+    // (commented out for now because we haven't ported everything yet)
+    // assert(canvas.getSaveCount() == 1);
+    return true;
   }
 
   void paintChild(RenderObject child, Point point) {
+    assert(debugCanPaintChild());
     final Offset offset = point.toOffset();
+    if (!child.requiresCompositing) {
+      child._paintWithContext(this, offset);
+    } else {
+      _compositeChild(child, offset, layer.parent, layer.nextSibling);
+    }
+  }
 
-    if (!child.requiresCompositing)
-      return child._paintWithContext(this, offset);
-    _compositChild(child, offset, layer.parent, layer.nextSibling);
+  void paintChildWithClip(RenderObject child, Point point, Rect clipRect) {
+    assert(debugCanPaintChild());
+    final Offset offset = point.toOffset();
+    if (!child.hasCompositedDescendant) {
+      // If none of the descendants require compositing, then we don't
+      // need to use a new layer here, because at no point will any of
+      // the children introduce a new layer of their own.
+      canvas.save();
+      canvas.clipRect(clipRect);
+      child._paintWithContext(this, offset);
+      canvas.restore();
+    } else {
+      // At least one of the descendants requires compositing. We
+      // therefore introduce a new layer to do the clipping, so that
+      // when the children are split into a new layer, the clip is not
+      // lost, as it would if we didn't introduce a new layer.
+      ClipLayer clip = new ClipLayer(offset: offset, clipRect: clipRect);
```

Comment:
As written, this will apply the offset twice.  You're passing the offset both via `point` and via `clipRect`.  IMHO, the best way to fix this is to have RenderClip pass in a clipRect that starts at Point.origin.